### PR TITLE
Fix tracking and improve draw call validation

### LINKED ISSUE #196: Avoid side effects in tracking iterators
The way `insert_barriers` is currently implemented, the iterators are produced for buffer and texture transitions. The iterators have side effects of *actually* updating the tracked state, and they are passed into gfx-hal.

Problem comes from the fact that gfx-hal doesn't guarantee that it will even touch the iterators, and the Metal backend simply drops the pipeline barriers. That discards any side effects the client hopes to get.

I don't think it's reasonable for gfx-hal to guarantee that all iterators are going to be iterated to the end. Instead, we should refactor wgpu-native code so that we don't rely on side effects.

This is a major correctness/security issue today, as our failure to track resources introduces subtle issues on the API/GPU side.

### PR DESCRIPTION
**Pull request by kvark**
Fixes #196 

## Validation
Note: the most difficult bit is still missing - we need to know the maximum index in an index buffer. This is not meant to be done in this PR though, the logic is good enough on its own.

## Tracking
Implements a custom iterator with Drop.
Also, reset tracking of resources when a command buffer is done.

### CODE REVIEW

File: wgpu-native/src/track.rs [hunk 1]
```diff
@@ -112,6 +112,45 @@ pub enum Stitch {
...
+                }
+            }
+        }
+    }
+}
+
+// Make sure to finish all side effects on drop
+impl<'a, I: TypedId, U: Copy + PartialEq> Drop for ConsumeIterator<'a, I, U> {
+    fn drop(&mut self) {
+        while self.next().is_some() {}
```
Comment (scottmcm) [inline wgpu-native/src/track.rs]: Canonical way to do this is `.for_each(drop)`

Comment (kvark) [reply wgpu-native/src/track.rs]: oh wow, fancy :)

Comment (scottmcm) [reply wgpu-native/src/track.rs]: References:
https://github.com/rust-lang/rust/commit/5c58eec0bd8cee8fb2a191396d5ad5b5c9b0116a
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect

File: wgpu-native/src/command/compute.rs [hunk 2]
```diff
@@ -144,34 +144,33 @@ pub extern "C" fn wgpu_compute_pass_set_pipeline(
...
-
-    let pipeline_layout_guard = HUB.pipeline_layouts.read();
-    let pipeline_layout = &pipeline_layout_guard[pipeline.layout_id];
-    let bind_group_guard = HUB.bind_groups.read();
-
-    pass.binder.pipeline_layout_id = Some(pipeline.layout_id.clone());
-    pass.binder
-        .reset_expectations(pipeline_layout.bind_group_layout_ids.len());
+    // Rebind resources
+    if pass.binder.pipeline_layout_id != Some(pipeline.layout_id.clone()) {
```
Comment (grovesNL) [inline wgpu-native/src/command/compute.rs]: How did the early out functionally change anything here? Or is this to make rebinding more explicit?

File: wgpu-native/src/track.rs (hunk 1 above)
Comment (grovesNL) [reply wgpu-native/src/track.rs]: Let's switch to `.for_each(drop)` here instead

File: wgpu-native/src/track.rs [hunk 3]
```diff
@@ -112,6 +112,45 @@ pub enum Stitch {
     Last,
 }
 
+pub struct ConsumeIterator<'a, I: TypedId, U: Copy + PartialEq> {
```
Comment (grovesNL) [inline wgpu-native/src/track.rs]: This looks nice :+1: I could see us using this in a few places in the future if we make it slightly more general (generic iterator and move the other fields into a generic context field)

File: wgpu-native/src/command/compute.rs (hunk 2 above)
Comment (kvark) [reply wgpu-native/src/command/compute.rs]: the compute side was fine, but I refactored it to be in line with the render side changes

**Comment by kvark** [comment]
Thanks for the review! All concerns have been addressed.
Note the last commit that is going in unreviewed, should be relatively safe.
bors r=grovesNL

**Pull request merged**